server/src/main/java/org/elasticsearch/index/query/first/FirstQuery.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.index.query.first;
+
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.Matches;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.ScorerSupplier;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.Scorer;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Set;
+
+
+public class FirstQuery extends Query {
+    private final Query[] queries;
+
+    public FirstQuery(Query[] queries) {
+        this.queries = queries;
+    }
+
+    public Query[] getQueries() {
+        return queries;
+    }
+
+    @Override
+    public Query rewrite(IndexReader reader) throws IOException {
+        if (queries.length == 1) {
+            return queries[0];
```

Comment:
I'm curious why in this case the query isn't rewritten? Is there something special about this case?

Reply:
I have copied this behaviour from Lucene's [DisjunctionMaxQuery](https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/search/DisjunctionMaxQuery.java#L203) and [BooleanQuery](https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java#L245),  but overall looks like `IndexSearcher` will [keep rewriting query in a loop](https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java#L756-L757) until it is the same, so this internal query `queries[0]` will be rewritten as well.